### PR TITLE
scroll virtualization

### PR DESCRIPTION
**Pull request by bhoomika08**

### CODE REVIEW

File: src/flat_list.jsx [hunk 1]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
   }
 }
 
 FlatList.defaultProps = {
   batchCount: 50,
   itemKey: "id",
-  isVirtual: false
+  isVirtual: false,
+  isScroll: false,
+  ListItemType: "span",
```
Comment (gash48) [inline src/flat_list.jsx]: should be separatorType not listItemType. And Use Camel Case

File: src/flat_list.jsx [hunk 2]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
+  itemToScrollPosition: null,
```
Comment (gash48) [inline src/flat_list.jsx]: Better name please. Should be 'positionToScroll'

File: src/flat_list.jsx [hunk 3]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
+  onScrollEnd: () => {}
```
Comment (gash48) [inline src/flat_list.jsx]: This represents scrollEnd in general. Should be onScrollToElementEnd

File: src/flat_list.jsx [hunk 4]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
+      </React.Fragment>
     ));
   }
 }
 
 FlatList.defaultProps = {
   batchCount: 50,
   itemKey: "id",
-  isVirtual: false
+  isVirtual: false,
```
Comment (gash48) [inline src/flat_list.jsx]: Better Names for these Props. Could be autoLoad, autoControlled, or u can think of better

File: src/flat_list.jsx [hunk 5]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
+  isScroll: false,
```
Comment (gash48) [inline src/flat_list.jsx]: should be "loadByScroll" or 'scrollControlled'

File: example/src/app/component/flat_list.jsx [hunk 6]
```diff
@@ -13,57 +15,116 @@ class FlatList extends React.Component {
...
-      <ItemRenderer
-        key={`List-${typeof item == "object" ? item[itemKey] : item}`}
-        item={item}
-        renderItem={renderItem}
-      />
+    const lastIndex = items.length - 1;
+    return (isVirtual || isScroll ? items : data).map((item, idx) => (
+      <React.Fragment key={`List-${typeof item == "object" ? item[itemKey] : item}`}>
+        <ItemRenderer item={item} renderItem={renderItem} />
+        <ListItemType ref={lastIndex == idx ? this.lastItemRef : `Ref-${idx}`}></ListItemType>
```
Comment (gash48) [inline example/src/app/component/flat_list.jsx:114]: Why do we need a condition now ?

File: src/flat_list.jsx [hunk 7]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
+    return (isVirtual || isScroll ? items : data).map((item, idx) => (
+      <React.Fragment key={`List-${typeof item == "object" ? item[itemKey] : item}`}>
+        <ItemRenderer item={item} renderItem={renderItem} />
+        <ListItemType ref={lastIndex == idx ? this.lastItemRef : `Ref-${idx}`}></ListItemType>
+      </React.Fragment>
     ));
   }
 }
 
 FlatList.defaultProps = {
```
Comment (gash48) [inline src/flat_list.jsx:150]: Be Sure To Add a thorough & descriptive info in ure readme regarding these props. And its usage

File: src/flat_list.jsx [hunk 8]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
-      <ItemRenderer
-        key={`List-${typeof item == "object" ? item[itemKey] : item}`}
-        item={item}
-        renderItem={renderItem}
-      />
+    const lastIndex = items.length - 1;
+    return (isVirtual || isScroll ? items : data).map((item, idx) => (
+      <React.Fragment key={`List-${typeof item == "object" ? item[itemKey] : item}`}>
+        <ItemRenderer item={item} renderItem={renderItem} />
+        <ListItemType ref={lastIndex == idx ? this.lastItemRef : `Ref-${idx}`}></ListItemType>
```
Comment (gash48) [inline src/flat_list.jsx]: Why Do We need a condition Now for last item in list ? All the list items will have this ref.
Please Use Better ref prefix. (`Ref-` => `separator-ref`) ..... and html attr & values are all small case.

Comment (gash48) [reply src/flat_list.jsx]: Instead of idx why cant we use `{typeof item == "object" ? item[itemKey] : item}`

Comment (gash48) [reply src/flat_list.jsx]: Depending on index is quite fishy in react

File: src/flat_list.jsx [hunk 9]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
-    const { data, renderItem, itemKey, isVirtual } = this.props;
+    const { data, renderItem, itemKey, isVirtual, isScroll, ListItemType } = this.props;
     const { items } = this.state;
-    return (isVirtual ? items : data).map(item => (
-      <ItemRenderer
-        key={`List-${typeof item == "object" ? item[itemKey] : item}`}
-        item={item}
-        renderItem={renderItem}
-      />
+    const lastIndex = items.length - 1;
```
Comment (gash48) [inline src/flat_list.jsx]: No Need

File: src/flat_list.jsx [hunk 10]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
+  scrollToElement(itemPosition) {
+    const { onScrollEnd } = this.props;
+    const { items } = this.state;
+    const referenceElement = itemPosition == items.length - 1 ? this.lastItemRef.current : this.refs[`Ref-${itemPosition}`];
+    clearTimeout(this.scrollEnd);
+    referenceElement.scrollIntoView({
+      behavior: "smooth",
+      block: "center"
+    });
+    this.scrollEnd = setTimeout(() => onScrollEnd(), 3000);
```
Comment (gash48) [inline src/flat_list.jsx]: `setTimeout(onScrollEnd, 3000)` might also work. Please Check this once

File: src/flat_list.jsx [hunk 11]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
+          items: [...items, ...data.slice(prevState.items.length, prevState.items.length + batchCount)]
+        }));
+      }, timeoutMilliSec);
+    }
+  }
+
+  scrollToElement(itemPosition) {
+    const { onScrollEnd } = this.props;
+    const { items } = this.state;
+    const referenceElement = itemPosition == items.length - 1 ? this.lastItemRef.current : this.refs[`Ref-${itemPosition}`];
```
Comment (gash48) [inline src/flat_list.jsx]: instead of position it will be itemKey or itemId

File: src/flat_list.jsx (hunk 2 above)
Comment (gash48) [reply src/flat_list.jsx]: Or keyToScroll

File: src/flat_list.jsx [hunk 12]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
+
+  componentWillUnmount() {
+    window.removeEventListener("scroll", this.handleOnScroll);
+  }
+
+  loadMoreItems() {
+    const { data, batchCount } = this.props;
+    const { items } = this.state;
+    clearTimeout(this.timeoutId);
+    const hasMore =
```
Comment (gash48) [inline src/flat_list.jsx]: Wont this simply be `data.length - items.length > 0`

File: src/flat_list.jsx [hunk 13]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
     this.state = {
       items: []
     };
+    this.lastItemRef = React.createRef();
```
Comment (gash48) [inline src/flat_list.jsx]: No Need

File: src/flat_list.jsx [hunk 14]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
+    this.scrollToElement = this.scrollToElement.bind(this);
   }
 
   componentDidMount() {
-    const { data, batchCount, isVirtual } = this.props;
-    if(isVirtual) {
+    const { data, batchCount, isVirtual, isScroll, scrollParent } = this.props;
+    const { items } = this.state;
+
+    if (isVirtual) {
```
Comment (gash48) [inline src/flat_list.jsx]: Both these if & else-if will be performed if data.length > 0

File: src/flat_list.jsx [hunk 15]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
+      if (oldData.length == data.length && itemChecker({ data: oldData }, { data })) {
+        if (items.length == 0)
+          this.setState({
+            items: data.slice(0, batchCount)
+          });
       } else {
         this.setState({ items: [] });
       }
     }
+    if (itemToScrollPosition >= 0 && itemToScrollPosition != null) {
```
Comment (gash48) [inline src/flat_list.jsx]: Comment Here

File: src/flat_list.jsx [hunk 16]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
-            this.setState(prevState => ({
-              items: [...items, ...data.slice(prevState.items.length, prevState.items.length + batchCount)]
-            }));
-          }, timeoutMilliSec);
-        }
+        this.loadMoreItems();
+      } else {
+        this.setState({ items: [] });
+      }
+    } else if (isScroll) {
```
Comment (gash48) [inline src/flat_list.jsx]: Comment here

File: src/flat_list.jsx [hunk 17]
```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
...
+      });
     }
   }
 
   componentDidUpdate({ data: oldData }) {
-    const { data, batchCount, isVirtual } = this.props;
+    const { data, batchCount, isVirtual, isScroll, itemToScrollPosition } = this.props;
     const { items } = this.state;
-
     if (isVirtual) {
```
Comment (gash48) [inline src/flat_list.jsx]: Comment Here

File: src/flat_list.jsx (hunk 15 above)
Comment (gash48) [reply src/flat_list.jsx]: Null check should preceed with >=0 check

**Comment by bhoomika08** [comment]
Done all CR-changes

File: src/flat_list.jsx [hunk 18]
```diff
@@ -1,71 +1,154 @@
...
+const timeoutMilliSec = 0;
 
 const itemChecker = (prevProps, nextProps) => {
   return isEqual(prevProps, nextProps);
 };
 
-const ItemRenderer = React.memo(({ item, renderItem }) => renderItem(item), itemChecker);
+class ItemRenderer extends React.Component {
+  shouldComponentUpdate() {
+    return !itemChecker();
```
Comment (gash48) [inline src/flat_list.jsx]: Not sending prevProps & nextProps in itemChecker ?

File: src/flat_list.jsx [hunk 19]
```diff
@@ -1,71 +1,154 @@
...
+}
 
 class FlatList extends React.Component {
   constructor(props) {
     super(props);
     this.state = {
       items: []
     };
+    this.getLastElementHeight = this.getLastElementHeight.bind(this);
+    this.getKey = this.getKey.bind(this);
```
Comment (gash48) [inline src/flat_list.jsx]: This Doesn't have to be a class Function. Can be just like 'itemChecker'

File: src/flat_list.jsx [hunk 20]
```diff
@@ -1,71 +1,154 @@
...
+  }
+}
 
 class FlatList extends React.Component {
   constructor(props) {
     super(props);
     this.state = {
       items: []
     };
+    this.getLastElementHeight = this.getLastElementHeight.bind(this);
```
Comment (gash48) [inline src/flat_list.jsx:29]: Same For This .... This Doesn't have to be a class Function

File: src/flat_list.jsx [hunk 21]
```diff
@@ -1,71 +1,154 @@
...
+    }
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener("scroll", this.handleOnScroll);
+  }
+
+  getLastElementHeight() {
+    const { items } = this.state;
+    return this.refs[`separator-ref-${this.getKey(items[items.length - 1])}`].getBoundingClientRect().bottom;
```
Comment (gash48) [inline src/flat_list.jsx]: What if ref is undefined or null ? Are u sure there will be a ref ?

File: src/flat_list.jsx [hunk 22]
```diff
@@ -1,71 +1,154 @@
...
   }
 
   componentDidUpdate({ data: oldData }) {
-    const { data, batchCount, isVirtual } = this.props;
+    const { data, batchCount, autoLoad, loadOnScroll, positionToScroll } = this.props;
     const { items } = this.state;
 
-    if (isVirtual) {
+    //---------- To Load Items Automatically -----------//
+    if (autoLoad) {
```
Comment (gash48) [inline src/flat_list.jsx]: Combine These Two If Else ... its possible

File: src/flat_list.jsx [hunk 23]
```diff
@@ -1,71 +1,154 @@
...
+        this.loadMoreItems();
+      } else {
+        this.setState({ items: [] });
+      }
+    }
+
+    //---------- To Load first batch of Items if items length is 0 before scrolling ---------//
+    else if (loadOnScroll) {
+      if (oldData.length == data.length && itemChecker({ data: oldData }, { data })) {
+        if (items.length == 0)
```
Comment (gash48) [inline src/flat_list.jsx]: This part has been used multiple times. Why Dont we extract to a common instance function ?

File: src/flat_list.jsx (hunk 20 above)
Comment (bhoomika08) [reply src/flat_list.jsx:29]: No I won't do it. It's not feasible.

File: README.md [hunk 24]
```diff
@@ -22,18 +28,26 @@ or with yarn:
 ```bash
 yarn add supercharged-list
 ```
+
```
Comment (gash48) [inline README.md:31]: Add the npm option too

File: README.md [hunk 25]
```diff
@@ -1,15 +1,21 @@
...
-So here comes the supercharged-list to overcome this situation. In this it will perform the diff checker of the previous and next props object. It will render only if there is diffrence in the content of object. This will not re-render the entire list with props as object.
+## What's this all About ?
+
+A memoized component which means that it will not re-render if props remain shallow-equal. But here come a twist, if the prop in list is an object and its content remain same, then also it will re-render as it will not remain shallow-equal due to different memory assigned to objects. So here comes the supercharged-list to overcome this situation. In this it will perform the diff checker of the previous and next props object. It will render only if there is diffrence in the content of object. This will not re-render the entire list with props as object.
+
+The List could be rendered in either of 3 ways:
```
Comment (gash48) [inline README.md]: Make This more emphatic. Bolder Important

File: README.md [hunk 26]
```diff
@@ -1,15 +1,21 @@
...
 
-Features: 
-1. Diff checker to render only those list items whose props object content is changed.
-2. For long lists, the items could render in batches automatically.
+1. Entire List to be rendered in one go. (autoLoad=false && loadOnScroll=false)
```
Comment (gash48) [inline README.md]: Also Mention this is default behaviour

File: README.md [hunk 27]
```diff
@@ -59,25 +73,24 @@ class Example extends React.Component {
...
       </div>
     );
   }
 }
-
 ```
-## More To Come
-
 
 ## Browsers support
```
Comment (gash48) [inline README.md]: No need for this section. Just Add React Dependency & Versions required

File: example/src/app/component/flat_list.jsx [hunk 28]
```diff
@@ -1,69 +1,162 @@
...
 const itemChecker = (prevProps, nextProps) => {
   return isEqual(prevProps, nextProps);
 };
 
-const ItemRenderer = React.memo(({ item, renderItem }) => renderItem(item), itemChecker);
+const getKey = (item, itemKey) => {
+  return typeof item == "object" ? item[itemKey] : item;
+};
+
+class ItemRenderer extends React.Component {
```
Comment (gash48) [inline example/src/app/component/flat_list.jsx:12]: Excellent

File: example/src/app/component/flat_list.jsx [hunk 29]
```diff
@@ -1,69 +1,162 @@
...
 import isEqual from "lodash.isequal";
 
+const timeoutMilliSec = 0;
+
 const itemChecker = (prevProps, nextProps) => {
   return isEqual(prevProps, nextProps);
 };
 
-const ItemRenderer = React.memo(({ item, renderItem }) => renderItem(item), itemChecker);
+const getKey = (item, itemKey) => {
```
Comment (gash48) [inline example/src/app/component/flat_list.jsx]: make this one liner

File: example/src/app/component/flat_list.jsx [hunk 30]
```diff
@@ -1,69 +1,162 @@
...
+        this.getFirstBatchOfItems();
+      }
+    } else if (loadOnScroll) {
+      if (data.length > 0 && items.length == 0) {
+        this.getFirstBatchOfItems();
+      }
+      const parent = scrollParent ? document.querySelector(scrollParent) : window;
+      const parentHeight = scrollParent ? parent.clientHeight : window.innerHeight;
+      parent.addEventListener("scroll", () => {
+        if (this.getLastElementHeight() - 300 <= parentHeight) {
```
Comment (gash48) [inline example/src/app/component/flat_list.jsx]: make 300 as a constant offset

File: src/flat_list.jsx (hunk 20 above)
Comment (gash48) [reply src/flat_list.jsx:29]: Ok Sorry

File: src/flat_list.jsx [hunk 31]
```diff
@@ -1,71 +1,162 @@
 import React from "react";
 import isEqual from "lodash.isequal";
 
-const timeoutMilliSec = 10;
+const timeoutMilliSec = 0;
 
 const itemChecker = (prevProps, nextProps) => {
```
Comment (gash48) [inline src/flat_list.jsx]: Make this one liner

File: src/flat_list.jsx [hunk 32]
```diff
@@ -1,71 +1,162 @@
...
   return isEqual(prevProps, nextProps);
 };
 
-const ItemRenderer = React.memo(({ item, renderItem }) => renderItem(item), itemChecker);
+const getKey = (item, itemKey) => {
```
Comment (gash48) [inline src/flat_list.jsx]: same as above

File: src/flat_list.jsx [hunk 33]
```diff
@@ -1,71 +1,162 @@
...
+        this.getFirstBatchOfItems();
+      }
+    } else if (loadOnScroll) {
+      if (data.length > 0 && items.length == 0) {
+        this.getFirstBatchOfItems();
+      }
+      const parent = scrollParent ? document.querySelector(scrollParent) : window;
+      const parentHeight = scrollParent ? parent.clientHeight : window.innerHeight;
+      parent.addEventListener("scroll", () => {
+        if (this.getLastElementHeight() - 300 <= parentHeight) {
```
Comment (gash48) [inline src/flat_list.jsx]: Make 300 as constant like 'timeoutMilliSec'

File: src/flat_list.jsx [hunk 34]
```diff
@@ -1,71 +1,162 @@
...
+
+  scrollToItem(item) {
+    const { onScrollToElementEnd, itemKey } = this.props;
+    const referenceElement = this.refs[`separator-ref-${getKey(item, itemKey)}`];
+    clearTimeout(this.scrollEnd);
+    referenceElement.scrollIntoView({
+      behavior: "smooth",
+      block: "center"
+    });
+    this.scrollEnd = setTimeout(onScrollToElementEnd, 3000);
```
Comment (gash48) [inline src/flat_list.jsx]: make this 3000 as a constant too

**Review by gash48 — CHANGES_REQUESTED** [review]
Requested Minor Changes

**Review by gash48 — APPROVED** [review]
Seems Fine

**Pull request merged**